File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    productExceptSelf(nums) {
        const result = nums
        let maxProduct = 1
        let zeroCount = 0

        for (let num of nums) {
            if (num === 0) {
                zeroCount++
            } else {
                maxProduct *= num
            }
        }

        if (zeroCount > 1) {
            return result.fill(0, 0, nums.length)
        }
        for (let i = 0; i < nums.length; i++) {
            if (zeroCount > 0) {
                if (nums[i] === 0) {
                    result[i] = maxProduct
                } else {
                    result[i] = 0
                }
            } else {
                result[i] = maxProduct / nums[i]
            }
        }

        return result
    }
}

File: Data Structures & Algorithms/house-robber/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums[0], nums[1])
        
        for i in range(2, len(nums), 1):
            nums[i] = max(nums[i-1], nums[i-2] + nums[i])
            nums[i-1] = max(nums[i-2], nums[i-1])
        return max(nums[-2], nums[-1])